### PR TITLE
Inclusion of PolicyOR

### PR DESCRIPTION
**Pull request by APysnack**
_note: resubmitted pull request after modifying the branch name._

This pull request implements an optional policyOR functionality that allows kmyth to seal/unseal against 2 different values of PCR's. Please feel free to contact with questions.

A formal demonstration of the functionality can be viewed here:
https://www.youtube.com/watch?v=ZCJS-8w6P90

If it assists the reviewer of the pull request, there are also about 2.5 hours of videos (all of which have "Kmyth" in the title) that discuss the majority of the steps that were taken in trying to solve this. Please note that we're all new to this technology, and the videos were partly meant to be informative, so they might be a bit lengthy at times (I recommend watching with 2x speed if they are).

If time permits in the future, there are a few things that could be refactored, and the policyOR could potentially be extended to more than just 2 policies (a max of 8).

Please let us know of any changes that may need to be made.

### CODE REVIEW

File: include/cipher/aes_gcm.h [hunk 1]
```diff
@@ -52,7 +52,7 @@ int aes_gcm_encrypt(unsigned char *key,
                     size_t key_len,
                     unsigned char *inData,
                     size_t inData_len, unsigned char **outData,
-                    size_t * outData_len);
```
Comment (PircDef) [inline include/cipher/aes_gcm.h:55]: There are a lot of changes that are only white space changes. These should not be part of the PR.

**Comment by PircDef** [comment]
Items in the bin folder should not be part of the repository, every user will build these for themselves.

File: src/tpm/object_tools.c [hunk 2]
```diff
@@ -400,12 +400,11 @@ int create_kmyth_object(TSS2_SYS_CONTEXT * sapi_ctx,
               temp_handle);
 
     // Make the newly recreated SRK primary object persistent
-    TSS2_RC rc = Tss2_Sys_EvictControl(sapi_ctx,
-                                       TPM2_RH_OWNER,
-                                       temp_handle,
-                                       &createObjectCmdAuths,
-                                       object_dest_handle,
-                                       &createObjectRspAuths);
+    rc = Tss2_Sys_EvictControl(sapi_ctx,
```
Comment (APysnack) [inline src/tpm/object_tools.c:403]: TSS2_RC rc was already previously declared in the same scope

File: include/tpm/kmyth_seal_unseal_impl.h [hunk 3]
```diff
@@ -24,54 +24,61 @@
...
- *                            'private area' struct for the sealed data object
+ * @param[out] sdo_private       TPM 2.0 sized buffer to hold the returned
+ *                               'private area' struct for the sealed data object
  *
  * @return 0 on success, 1 on error
  */
 int tpm2_kmyth_seal_data(TSS2_SYS_CONTEXT * sapi_ctx,
                          uint8_t * sdo_data,
-                         int sdo_dataSize,
+                         size_t sdo_dataSize,
```
Comment (APysnack) [inline include/tpm/kmyth_seal_unseal_impl.h:71]: size_t was passed into function

File: src/tpm/kmyth_seal_unseal_impl.c [hunk 4]
```diff
@@ -28,21 +27,90 @@
  */
 extern const cipher_t cipher_list[];
 
+//############################################################################
```
Comment (bit-bender) [inline src/tpm/kmyth_seal_unseal_impl.c:30]: This function seems like more of a "formatting" or "TPM interface" utility. My preference would be to move it, say, to utils/src/formatting_tools.c.

File: src/tpm/kmyth_seal_unseal_impl.c [hunk 5]
```diff
@@ -28,21 +27,90 @@
...
+    ul = strtoul(substr, NULL, 16);
+    expectedPolicyBuffer[i / 2] = ul;
+  }
+
+  // converts the byte array into a TPM2B_DIGEST struct
+  memcpy(digest, expectedPolicyBuffer, sizeof(expectedPolicyBuffer));
+  return 0;
+}
+
+//############################################################################
```
Comment (bit-bender) [inline src/tpm/kmyth_seal_unseal_impl.c:64]: As the previous function, this on seems like more of a "formatting" or "TPM interface" utility. My preference would be to move it, say, to utils/src/formatting_tools.c.

File: src/tpm/marshalling_tools.c [hunk 6]
```diff
@@ -408,7 +596,15 @@ int create_ski_bytes(Ski input, uint8_t ** output, size_t * output_length)
...
-                         &wk_priv_data, &wk_priv_size, wk_priv_offset))
+                         &wk_priv_data,
+                         &wk_priv_size,
+                         wk_priv_offset,
+                         &input.policyBranch1,
+                         &p_branch_1_data,
+                         &p_branch_1_size,
+                         p_branch_1_offset,
+                         &input.policyBranch2,
+                         &p_branch_2_data, &p_branch_2_size, p_branch_2_offset))
```
Comment (bit-bender) [inline src/tpm/marshalling_tools.c:607]: In the error handling code for this call, I think we need to add free() calls for `p_branch_1_data` and `p_branch2_data`.

**Review by bit-bender — CHANGES_REQUESTED** [review]

**Comment by bit-bender** [comment]
The link, in the first comment, to the excellent video demonstrating this functionality is obsolete (no longer works). I think I found a newer version, but an updated link would be helpful.

**Comment by APysnack** [comment]
> The link, in the first comment, to the excellent video demonstrating this functionality is obsolete (no longer works). I think I found a newer version, but an updated link would be helpful.

Yes, for the original video the functionality that allows the user to pass the "-e" flag to retrieve the TPM2B_Digest policy associated with the user's "current" PCR values had not been written yet, so the new video shows that functionality. I have updated the link in the initial pull request. 

Thank you for the review. Once I get the chance, I will  look into each of these comments in depth and try to make changes per your requests.

File: src/tpm/kmyth_seal_unseal_impl.c [hunk 7]
```diff
@@ -28,21 +27,90 @@
...
+// convert_digest_to_string()
+//############################################################################
+static int convert_digest_to_string(TPM2B_DIGEST * digest, char *string_buf)
+{
+  // total number of hex values in the TPM2B digest
+  size_t digest_size = (8 * sizeof(digest)) + 2;
```
[the remark this answers is not in the document]
Comment (bit-belcher) [reply src/tpm/kmyth_seal_unseal_impl.c:70]: will have to address this on closer look - it is one of the things to do which is high priority

File: src/tpm/kmyth_seal_unseal_impl.c [hunk 8]
```diff
@@ -28,21 +27,90 @@
...
+
+  char hex_buf[digest_size];
+
+  memcpy(hex_buf, digest, digest_size);
```
[the remark this answers is not in the document]
Comment (bit-belcher) [reply src/tpm/kmyth_seal_unseal_impl.c:74]: address this (see above)

File: src/tpm/object_tools.c (hunk 2 above)
Comment (bit-belcher) [reply src/tpm/object_tools.c:403]: Does not matter. The rc (return code) is reused.

File: src/tpm/kmyth_seal_unseal_impl.c (hunk 4 above)
Comment (bit-belcher) [reply src/tpm/kmyth_seal_unseal_impl.c:30]: This was done.

File: src/tpm/kmyth_seal_unseal_impl.c [hunk 9]
```diff
@@ -28,21 +27,90 @@
...
+  //  buffer with values indicated in the string
+  for (size_t i = 0; i < strlength; i += 2)
+  {
+    strncpy(substr, &str[i], 2);
+    ul = strtoul(substr, NULL, 16);
+    expectedPolicyBuffer[i / 2] = ul;
+  }
+
+  // converts the byte array into a TPM2B_DIGEST struct
+  memcpy(digest, expectedPolicyBuffer, sizeof(expectedPolicyBuffer));
```
[the remark this answers is not in the document]
Comment (bit-belcher) [reply src/tpm/kmyth_seal_unseal_impl.c:60]: Used 8*sizeof(digest) + 2 in subsequent code.

**Review by bit-belcher — APPROVED** [review]
Noted a few changes did not make it into the final code state. I removed char s[n] declarations and relied on malloc in a few cases (convert_.. functions). Spacing changes are ok  and another main point of this commit is to change to dual policy cases with respective changes to the callers/callees.

**Review by bit-belcher — APPROVED** [review]
Noted a few changes did not make it into the final code state. I removed char s[n] declarations and relied on malloc in a few cases (convert_.. functions). Spacing changes are ok  and another main points of this commit is to change to dual policy cases with respective changes to the callers/callees and add another CLI argument. This commit also should've been separated into 3-4 commits since these differences were so numerous.

**Comment by bit-bender** [comment]
These proposed changes are addressed and/or incorporated in pull  request #166.
